### ISSUE TITLE
[svg-native] Consider supporting some subset of CSS

### ISSUE THREAD

**Issue by litherum** [w3c/svgwg#658]
We should determine whether or not we want to support some CSS in SVG Native.

There are many different ways to slice this:
- Which properties, which values
- `<style>` vs `style=`
- Media queries
- `@keyframes`
- Other at-rules
- CSS Variables (possibly a different policy for declaring new ones vs using existing ones)
  - Require CSS lexer on all values?
- Which selectors / combinators
- CSS global keywords `unset`, etc.
- `calc()`, `var()`, `env()`
- More?

**Comment by litherum** [comment]
@jarek-foksa [says](https://github.com/w3c/svgwg/issues/657#issuecomment-482832311):

> The "style" attribute should be allowed since the SVG 2 spec states that "(...) in the future, any new properties that apply to SVG content will not gain presentation attributes. Therefore, authors are suggested to use styling properties, either through inline ‘style’ properties or style sheets, rather than presentation attributes, for styling SVG content."

**Comment by dirkschulze** [comment]
...
* `!important`

**Comment by dirkschulze** [comment]
Proposal was to
* support the following presentation attributes/CSS properties:
    * `clip-path`
    * `clip-rule`
    * `color`
    * `display` (values other than `none` ignored)
    * `fill`
    * `fill-opacity`
    * `fill-rule`
    * `opacity`
    * `stroke`
    * `stroke-dasharray`
    * `stroke-dashoffset`
    * `stroke-linecap`
    * `stroke-linejoin`
    * `stroke-miterlimit`
    * `stroke-opacity`
    * `stroke-width`
    * `stop-color`
    * `stop-opacity`
    * `visibility` (values other than `visible`, `hidden` ignored)
* support the [`style` attribute](https://www.w3.org/TR/SVG2/styling.html#ElementSpecificStyling) with its declaration list.
* **No support** for `<style>` (and therefore) media queries, any @-rules, selectors or combinators
* Property inheritance stays.
* Values and units have the following restrictions
    * Pre-defined keywords (`inherit`, `unset`, `reset`,...), important declarations get ignored.
    * Only absolute length units or percentage.
    * functional notations limited to: `url()`, `rgb()` (and possibly `var()` with limitations... other issue). **No** `calc()`, `att()`, `image()`, `rgba()`, ... .
    * Color values other than color keywords (but system colors), `rgb()`, hex (3-6 digits) and `var()` (with strict limitations) get ignored.
    * Support of `currentColor` for `fill`, `stroke`, `stroke-color`.

**Comment by svgeesus** [comment]
Good to see some clarification here. As [I said in my initial review comments](https://lists.w3.org/Archives/Public/www-svg/2019Apr/0010.html), we need to be clear what exactly from CSS is supported, or not supported. In my initial comment, I said

> probably means that it removes CSS selectors, CSS stylesheets, and 
> cascading. CSS properties and inheritance are still used.

Having a defined whitelist of supported properties is an improvement.

Allowing the style attribute might be valuable, although in the absence of internal or external style sheets it adds nothing compared to just presentational attributes, and adds complexity (comments, multiple declarations, specificity if the presentational attribute is also specified).

I agree that internal and external stylesheets should be forbidden (and thus the `style` element), which also removes at-rules, selectors, and cascading.

Restrictions on CSS Values and Units is probably a good idea, although I am unsure why `rgba()` should be disallowed (and note that CSS Color 4 now allows an optional alpha term even for `rgb()`.)

**Comment by BigBadaboom** [comment]
> although in the absence of internal or external style sheets it adds nothing compared to just presentational attributes

Not exactly.  There are some properties that are not available as attributes.  For example, `shape-rendering`, which people may want to use for icons.

**Comment by dirkschulze** [comment]
@svgeesus The `style` attribute explicitly only allows a declaration list (a list of declarations). That limits it's CSS-ness a lot.

@BigBadaboom You should add a new issue for `shape-rendering` I don't think it is under consideration at the moment. Also, I am not sure if it even has an effect on todays web browsers.

**Comment by BigBadaboom** [comment]
@dirkschulze 

Ok. I have create #679 for that issue.

> Also, I am not sure if it even has an effect on todays web browsers.

It's supported by Chrome and Blink at least.

**Comment by svgeesus** [comment]
> Not exactly. There are some properties that are not available as attributes.

Oh, good point. Though a cleaner way would be to add any missing presentation attributes, especially ones that SVG Native needs and particularly so if any implementation already supports them.

**Comment by litherum** [comment]
Forked to https://github.com/w3c/svgwg/issues/681
